[[ libscript ]] Add Objective-C block support

livecodemichael opened the pull request:
This patch adds support for lcb scripts to use Objective-C blocks.
Two handlers have been added to the Objective-C library, allowing
for the creation and deletion of block pointers.

livecode-vulcan committed:
[[ libscript ]] Add Objective-C block support

This patch adds support for lcb scripts to use Objective-C blocks.
Two handlers have been added to the Objective-C library, allowing
for the creation and deletion of block pointers.

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+	end if
+end handler
+
+public foreign type BlockDescriptor binds to "MCAggregateTypeInfo:ffrrr"
```
I suggest `_ObjcBlockDescriptor` - they have to be public at the moment due to an LCB bug - so using a `_` prefix at least gives some modicum of indication that they are actually meant to be private/internal.

runrevmark replied:
Order wise I suggest the same as above - constants then types than foreign handlers all as one 'block'. I think

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+end handler
+
+public foreign type BlockDescriptor binds to "MCAggregateTypeInfo:ffrrr"
+public foreign type Block binds to "MCAggregateTypeInfo:rEErr"
```
I suggest `_ObjcBlock`

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+public foreign type BlockDescriptor binds to "MCAggregateTypeInfo:ffrrr"
+public foreign type Block binds to "MCAggregateTypeInfo:rEErr"
+type BlockPointer is Pointer
```
I suggest `ObjcBlockPointer`

runrevmark replied:
It also needs to be public I think

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+private constant kBlockSize is 32
+private constant kBlockDescriptorSize is 40
+
+public handler CreateBlockPointerFromHandler(in pHandler as any, out rBlockPtr as optional BlockPointer) returns Boolean
```
`CreateObjcBlockPointerFromHandler`

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+	end unsafe
+end handler
+
+public handler DeleteBlockPointer(in pBlockPointer as BlockPointer)
```
`DeleteObjcBlockPointer`

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
 
+/* Block handling */
+
+type LibraryHandle is Pointer
```
I suggest:
```
private variable ...

private constant ...
private type LibraryHandle ...
private foreign handler(s) ...

... EnsureNSConcreteGlobalBlock definition
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+private foreign handler GetSymbolPointer(in pHandle as LibraryHandle, in pSymbol as ZStringUTF8) returns optional Pointer binds to "dlsym"
+private foreign handler CloseLibrary(in pHandle as LibraryHandle) returns CSInt binds to "dlclose"
+
+private unsafe handler EnsureNSConcreteGlobalBlock() returns nothing
```
```
/* Fetch the `isa` pointer for the (internal) Obj-C concreate block class */
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+		return
+	end if
+
+	variable tSystem as LibraryHandle
```
```
/* As we cannot bind to the pointer to a symbol as a variable, we use
 * dynamic library loading and resolution functions to do so directly. */
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+private variable mNSConcreteGlobalBlock as optional Pointer
+
+constant kRTLD_LAZY is 1
+private foreign handler OpenLibrary(in pPath as ZStringUTF8, in pFlag as CSInt) returns optional LibraryHandle binds to "dlopen"
```
Break the foreign handler definitions as we have above - i.e. continuation to newline before the `binds`

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+foreign handler MCHandlerGetFunctionPtr(in pHandler as any, out rPtr as Pointer) returns CBool binds to "<builtin>"
+
+private foreign handler CopyBlockDescriptor(inout pBlock as BlockDescriptor, in pSize as UIntSize, out rNewBlock as Pointer) returns CBool binds to "MCMemoryAllocateCopy"
```
Continue the `binds` on the new line for each foreign handler

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+private foreign handler MCMemoryDelete(in pBlock as Pointer) returns nothing binds to "<builtin>"
+
+private foreign handler BlockPointerToBlock(out rBlock as Block, \
```
We haven't continued args elsewhere, so retract that and just continue the `binds`

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+public handler CreateBlockPointerFromHandler(in pHandler as any, out rBlockPtr as optional BlockPointer) returns Boolean
+	unsafe
+		EnsureNSConcreteGlobalBlock()
```
```
/* Make sure the `isa` pointer has been initialized */
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+	unsafe
+		EnsureNSConcreteGlobalBlock()
+
+		variable tFunctionPtr as Pointer
```
```
/* Fetch the (C) function pointer for the given handler - this is used later when
 * constructing the block's record and is called when the block is invoked. */

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+			return false
+		end if
+
+		variable tBlock as Block
```
Move this var def to above where it is defined.

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+		end if
+
+		variable tBlock as Block
+		variable tBlockDescriptor as BlockDescriptor
```
```
/* Initialize the list of block descriptor aggregate members. The first field is the
 * flags where 0 means global block; the second field is the size of the block
 * record; the other members are not used (for our purposes). */
```
```

Remove newline after this var def.

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+		put [0, kBlockSize, nothing, nothing, nothing] into tBlockDescriptor
+
+		variable tBlockDescriptorPointer as Pointer
```
```
/* Copy the block descriptor after being (implicitly) converted to an aggregate
 * into a heap allocated memory block. */
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+			return false
+		end if
+
+		put [mNSConcreteGlobalBlock, 0x30000000, 0, tFunctionPtr, tBlockDescriptorPointer] into tBlock
```
```
/* Initialize the list of block members. The first field is the obj-c objects 'isa' pointer; the
 * second is the flags word; the third is unused for our purposes; the fourth is the (C)
 * function pointer it wraps; and the fifth is the block descriptor (which mainly defines
 * the actual size of the block - as we don't add any block variables, this is a fixed size. */
variable ...
put ...
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+		put [mNSConcreteGlobalBlock, 0x30000000, 0, tFunctionPtr, tBlockDescriptorPointer] into tBlock
+
+		variable tBlockPtr as BlockPointer
```
```
/* Copy the block after being (implicitly) converted to an aggregate into a heap
 * allocated memory block. */
```

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+		end if
+
+		put tBlockPtr into rBlockPtr
+		return true
```
Move this to outside the `unsafe` block.

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+
+public handler DeleteBlockPointer(in pBlockPointer as BlockPointer)
+	unsafe
+		variable tBlock as Block
```
```
/* Unpack the pointer to block aggregate as an aggregate. */

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+	unsafe
+		variable tBlock as Block
+		BlockPointerToBlock(tBlock, pBlockPointer, kBlockSize)
+		MCMemoryDelete(tBlock[5])
```
```
/* Delete the block's descriptor (fifth field). */

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,89 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+		variable tBlock as Block
+		BlockPointerToBlock(tBlock, pBlockPointer, kBlockSize)
+		MCMemoryDelete(tBlock[5])
+		MCMemoryDelete(pBlockPointer)
```
```
/* Delete the block itself. */
```

runrevmark reviewed (changes requested):
Okay so this is my suggestions for code layout / comments - I think the comments on what the code is doing make sense... It matches my recollection of the saga getting it to work on Gitter anyway! Any improvements/thoughts more than welcome :)

runrevmark commented:
This needs a release note `feature-objc_blocks.md` perhaps along the lines of the existing `feature-objc_delegates.md`.

runrevmark commented:
@livecodemichael I can't get this to compile as it is - the `objc.lcb` file is compiled in the bootstrap phase as it is currently a 'syntax' module (meaning it adds syntax to lcb - even though it doesn't). This means it can only use core syntax (control structures and method calls). The new code uses a few other things though, which would be hard to work around. (Currently you get a syntax error when doing a full project rebuild - when building the bootstrap lc-compile).

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,208 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+parameter of its signature being the block pointer. The remaining parameters
+should match those of the Objective-C block.
+
+References:
```
I suggest making it clear that the ObjcBlockPointer must be deleted when no longer used. e.g.
```
The lifetime of a created `ObjcBlockPointer` is not automatically managed. When such a value
has no more references to it and it is no longer going to be used, `DeleteObjcBlockPointer`
should be used to free the resources used by it.
```

runrevmark commented:
@livecodemichael Okay so the objc.lcb file should be moved to the 'other' list of module files (next to unittest_IMPL) that fixes the compile-lcb-error.

However, it then uncovers another issue - aggregate type definitions should not be shimmed (this happens for all builtin modules, type defs and handlers have explicit C++ shims built for them so that they never have to be dlsym'd at runtime). Currently we do create shims for parameterized foreign types (those with ':' in the binding string) which causes a compile error - but the shims would never be used (as the module loading code does different things for foreign types which contain ':' compared to ones which don't).

So, turns out this is easy to fix - so can be done in a commit in this PR.

In emit.cpp there is a method EmittedBuiltinAdd which checks to see if a method should be added as a shimmed builtin, and if so does so. All we need to do is make sure parameterized foreign types are not processed in this way.

Something along the lines of:
```
    if (!p_force && (!OutputFileAsC || OutputFileAsAuxC))
    {
        return MCNameGetString(to_mcnameref(p_symbol_name));
    }

*    /* If this is a parameterized foriegn type, then don't shim it */
+    if (p_type_index == UINDEX_MAX &&
+        strchr(cstring_from_nameref(p_symbol_name), ':') != NULL)
+    {
+        return MCNameGetString(to_mcnameref(p_symbol_name));
+    }
 ```

Here p_type_index is UINDEX_MAX if the symbol is a foreign type, and p_symbol_name is the binding string - so we just have to check the latter for ':' to tell if it parameterized.

livecode-vulcan committed:
[[ toolchain ]] Do not shim aggregate type definitions

This patch makes sure we do not generate shims for aggregate type
definitions. Currently, we create shims for parameterised foreign
types (those with a ":" in the binding string). These shims are
never used and cause compiler errors when using aggregates in
builtin modules.

livecode-vulcan committed:
[[ docs ]] Add release note for Objective-C block support

This patch adds a release note for Objective-C block support, where
handlers have been added tp the `objc` module facilitating the use
of Objective-C blocks.

runrevmark commented on docs/lcb/notes/feature-objc_blocks.md:
```diff
@@ -0,0 +1,79 @@
+argument and the variable into which the `ObjcBlockPointer` should be placed as
+the second.
+
+	private variable sRequestPermissionsCompletionHandler as optional ObjcBlockPointer
```
Does this need to be in a triple-backquote block (I presume our markdown parser handles that...)?

livecodemichael replied:
I've followed a similar structure to what we have in https://github.com/livecode/livecode/blob/develop/docs/lcb/notes/feature-objc_delegates.md which uses indents for code blocks. I guess that is maybe what the guide generator is expecting.

The file is rendered correctly by GitHub https://github.com/livecode/livecode/blob/develop/docs/lcb/notes/feature-objc_delegates.md

runrevmark commented on docs/lcb/notes/feature-objc_blocks.md:
```diff
@@ -0,0 +1,79 @@
+require blocks, passing the created `ObjcBlockPointer` where an Objective-C
+block would be expected.
+
+	private foreign handler ObjC_AVCaptureDeviceRequestAccessForMediaType(in pMediaType as ObjcId, in pCompletionHandler as ObjcBlockPointer) \
```
Triple backquote?

runrevmark commented on docs/lcb/notes/feature-objc_blocks.md:
```diff
@@ -0,0 +1,79 @@
+such a value has no more references to it and it is no longer going to be used,
+`DeleteObjcBlockPointer` should be used to free the resources used by it.
+
+	public handler AudioLibraryFinalize()
```
Triple backquote?

runrevmark commented on libscript/src/objc.lcb:
```diff
@@ -610,4 +610,212 @@ public handler CreateObjcInformalDelegateWithContext(in pProtocol as List, in pH
+/**
+Create an Objective-C block pointer that wraps an LCB handler.
+
+Example:
```
I take it 'example' blocks in the docs comments are automagically made code-like in whatever output they are generated to?

livecodemichael replied:
Having now got the dictionary to build correctly with this patch, I can confirm they are automatically formatted as code blocks : )

runrevmark reviewed (changes requested):
Could you sink the toolchain patch below the objc patch as compiling will fail if they are the other way round and you checkout the lower one currently (which is important for git bisection!).

livecode-vulcan committed:
[[ docs ]] Handle public type definitions in the parser

This patch updates the doc parser so that it correctly handles
definitions of the form `public type ...`.

runrevmark commented:
@livecode-vulcan review ok 9beeb1c

livecode-vulcan commented:
💙 review by @runrevmark ok 9beeb1c6f1f2b1aa9896a352ebb587e49c5fb6c2

livecode-vulcan commented:
😎 test success 9beeb1c6f1f2b1aa9896a352ebb587e49c5fb6c2

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/738)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/147)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/147)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/147)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/147)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2264)
* try-community-universal-ios-iphoneos13.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.2/builds/39)
* try-community-universal-ios-iphonesimulator13.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.2/builds/39)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2293)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2280)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2279)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2313)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2253)
